driver-core/src/main/com/mongodb/client/model/expressions/Expressions.java

```diff
@@ -0,0 +1,217 @@
+        for (double n : array) {
+            list.add(new BsonDouble(n));
+        }
+        return new MqlExpression<>((cr) -> new AstPlaceholder(new BsonArray(list)));
+    }
+
+    public static NumberExpression of(final Decimal128 of) {
+        Assertions.notNull("Decimal128", of);
+        return new MqlExpression<>((codecRegistry) -> new AstPlaceholder(new BsonDecimal128(of)));
+    }
+
+    public static ArrayExpression<NumberExpression> ofNumberArray(final Decimal128... array) {
+        Assertions.notNull("array", array);
+        List<BsonValue> result = new ArrayList<>();
+        for (Decimal128 e : array) {
+            Assertions.notNull("elements of array", e);
+            result.add(new BsonDecimal128(e));
+        }
+        return new MqlExpression<>((cr) -> new AstPlaceholder(new BsonArray(result)));
+    }
+
+    public static DateExpression of(final Instant of) {
+        Assertions.notNull("Instant", of);
+        return new MqlExpression<>((codecRegistry) -> new AstPlaceholder(new BsonDateTime(of.toEpochMilli())));
+    }
+
+    public static ArrayExpression<DateExpression> ofDateArray(final Instant... array) {
+        Assertions.notNull("array", array);
+        List<BsonValue> result = new ArrayList<>();
+        for (Instant e : array) {
+            Assertions.notNull("elements of array", e);
+            result.add(new BsonDateTime(e.toEpochMilli()));
+        }
+        return new MqlExpression<>((cr) -> new AstPlaceholder(new BsonArray(result)));
+    }
+
+    /**
+     * Returns an expression having the same string value as the provided
+     * string.
+     *
+     * @param of the string
+     * @return the string expression
+     */
+    public static StringExpression of(final String of) {
+        Assertions.notNull("String", of);
+        return new MqlExpression<>((codecRegistry) -> new AstPlaceholder(new BsonString(of)));
+    }
+
+
+    public static ArrayExpression<StringExpression> ofStringArray(final String... array) {
+        Assertions.notNull("array", array);
+        List<BsonValue> result = new ArrayList<>();
+        for (String e : array) {
+            Assertions.notNull("elements of array", e);
+            result.add(new BsonString(e));
+        }
+        return new MqlExpression<>((cr) -> new AstPlaceholder(new BsonArray(result)));
+    }
+
+    @SafeVarargs  // nothing is stored in the array
+    public static <T extends Expression> ArrayExpression<T> ofArray(final T... array) {
+        Assertions.notNull("array", array);
+        return new MqlExpression<>((cr) -> {
+            List<BsonValue> list = new ArrayList<>();
+            for (T v : array) {
+                Assertions.notNull("elements of array", v);
+                list.add(((MqlExpression<?>) v).toBsonValue(cr));
+            }
+            return new AstPlaceholder(new BsonArray(list));
+        });
+    }
+
+    public static DocumentExpression of(final Bson document) {
```

Comment:
To differentiate between `of(Bson)` producing `DocumentExpression` and `of(Bson)` producing `MapExpression`, we call the latter `ofMap` and have:
```
DocumentExpression of(Bson);
MapExpression ofMap(Bson);
MapExpression ofMap();

```

Renaming `of(Bson)` to `ofDocument(Bson)` appears to make naming more consistent:

- if `of` can be distinguished based on the input type, we use the name "of";
- if `of` cannot be distinguished based on the input type, we include the result type in the name.

Currently all methods follow the above rules expect for `DocumentExpression of(Bson)`.

Reply:
Unsure. It seems that ofDocument is a "default" that would be used much more often than map, and that this is similar to how we handled current/currentAsMap?

Reply:
This is indeed consistent with `current/currentAsMap`.